### ISSUE TITLE
For loop breaks when freeing a node that's a key in a dictionary while iterating over said dictionary.

### ISSUE THREAD

**Issue by CritCorsac** [godotengine/godot#40608]
**Godot version:**
3.2.2 and 3.2.3 beta1

**Issue description:**
While using a for loop to iterate through the contents of a dictionary, if one of the dictionaries keys is a node and said node gets freed the loop will break unexpectedly. This does not occur in 3.2.1 as the loop continues as expected in that version.

**Steps to reproduce:**
Simply add this code to a node that's a parent to multiple other nodes. The second for loop will only loop once then break.
```
var test_dictionary = {}

func _ready():
	for a in get_children():
		test_dictionary[a] = "useless_value"
	for a in test_dictionary:
		a.free()
		print("looped")

```
Interestingly, calling queue_free() instead of free() causes no problems.

**Minimal reproduction project:**
[LoopBreakBug.zip](https://github.com/godotengine/godot/files/4962175/LoopBreakBug.zip)

**Comment by barbaros83** [comment]
i think you are supposed to use queue_free()

**Comment by CritCorsac** [comment]
I think queue_free() should be used over free() whenever possible but there are a few occasions when you need to use free() instead. For example, if you need an up to date get_child_count() within the same frame the nodes were freed instead of having to wait for the end of it.

**Comment by Calinou** [comment]
I believe this may have changed as a result of the dangling Variant fix. Can you test this behavior both in debug and release builds (**Export with Debug** unchecked)?

**Comment by akien-mga** [comment]
Seems fixed in 3.2.3-stable and current `3.2` (5357e6a974cadf58bfdfa215250bdc12dd5fdea5).
